File: Python/Lesson_191_n_queens_ii.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 191 -- N Queens II
# Category   : Backtracking
# Difficulty : Hard
# Study Plan : Day 96
# =============================================================
#
# QUESTION:
#   Return number of distinct solutions for n-queens.
# =============================================================
def totalNQueens(n):
    cnt=[0]; cols=set(); d1=set(); d2=set()
    def bt(r):
        if r==n: cnt[0]+=1; return
        for c in range(n):
            if c in cols or r-c in d1 or r+c in d2: continue
            cols.add(c); d1.add(r-c); d2.add(r+c); bt(r+1)
            cols.remove(c); d1.remove(r-c); d2.remove(r+c)
    bt(0); return cnt[0]

if __name__=="__main__":
    print(totalNQueens(4))
    print(totalNQueens(8))

File: CPP/Lesson_191_n_queens_ii.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 191 -- N Queens II
// Category   : Backtracking
// Difficulty : Hard
// Study Plan : Day 96
// =============================================================
//
// QUESTION:
//   Return number of distinct solutions for n-queens.
// =============================================================
#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
int CNT,N; set<int> C,D1,D2;
void bt(int r){if(r==N){CNT++;return;}for(int i=0;i<N;i++){if(C.count(i)||D1.count(r-i)||D2.count(r+i))continue;C.insert(i);D1.insert(r-i);D2.insert(r+i);bt(r+1);C.erase(i);D1.erase(r-i);D2.erase(r+i);}}
int totalNQueens(int n){CNT=0;N=n;C.clear();D1.clear();D2.clear();bt(0);return CNT;}
int main(){cout<<totalNQueens(4)<<"\n"<<totalNQueens(8)<<"\n";}
